### PR TITLE
I18N: Replace `@wordpress/i18n` with Tannin

### LINKED ISSUE #8577: I18N: Use Tannin instead of `@wordpress/i18n`
## Feature Description

The story editor and dashboard use `@web-stories-wp/i18n`, which currently is a simple wrapper around `@wordpress/i18n` plus some useful utility functions.

We did this in #6183 in order to remove a direct dependency on a GPL-licensed (but externalized!) dependency across tons of files, so that we could eventually swap out `@wordpress/i18n` for something else.

It's something we _have to_ to before being able to distribute the editor & dashboard package without this dependency being externalized.

And this "something else" is [Tannin](https://github.com/aduth/tannin), the library that also powers `@wordpress/i18n` but is MIT-licensed.

Since Tannin essentially offers the same API as `@wordpress/i18n`, no changes to `story-editor` & `dashboard` need to be made.

## Alternatives Considered

* Just wait for the Gutenberg package to be dual-licensed.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

In `@web-stories-wp/i18n`:

Instantiate a `new Tannin` instance. Export `__`, `_x`, etc. helper functions acting as wrappers around `dcnpgettext`, so that the API stays the same for consumers. Also export `sprintf` from `@tannin/sprintf`. Perhaps also export a `setLocaleData` method to set `tannin.data` from outside.

In `wp-story-editor` and `wp-dashboard`:

Option A:

Get locale data from `@wordpress/i18n` using `getLocaleData` and pass this to our Tannin instance.

Unfortunately `getLocaleData` is only available as of very recently and does not exist on WP 5.5.

Option B:

Don't use `wp_set_script_translations` but use something custom [like Site Kit does](https://github.com/google/site-kit-wp/blob/d6ea05554e20353df9e770abd193c98b006cdf79/includes/Core/Assets/Script.php) to pass locale data to our JavaScript via some inline script.

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

See #8577 for context

## Summary

Replace `@wordpress/i18n` with Tannin, the underlying library used by it.

## To-do

- [x] Ensure string extraction still works (i18n lint should pass)
- [x] Reduce bundle size if possible
- [x] Ensure  that `wp-i18n` script is not unnecessarily enqueued
- [x] Fix inline script calls for all scripts _and_ chunks

## User-facing changes

N/A

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Switch your site/user language to something like Portuguese (Brazil)
2. Go to Dashboard -> Updates to ensure all translations are installed
3. Go to the story dashboard and the story editor and verify that the majority of strings are translated.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8577

### CODE REVIEW

File: includes/Admin/Editor.php [hunk 1]
```diff
@@ -340,6 +340,7 @@ public function get_editor_settings(): array {
 				'encodeMarkup'                 => $this->decoder->supports_decoding(),
 				'metaBoxes'                    => $this->meta_boxes->get_meta_boxes_per_location(),
 				'ffmpegCoreUrl'                => trailingslashit( WEBSTORIES_CDN_URL ) . 'js/@ffmpeg/core@0.10.0/dist/ffmpeg-core.js',
+				'localeData'                   => json_decode( load_script_textdomain( self::SCRIPT_HANDLE, 'web-stories' ), true ),
```
Comment (spacedmonkey) [inline includes/Admin/Editor.php]: Should this be moved to a trait or another class or something?

Comment (swissspidy) [reply includes/Admin/Editor.php]: Probably not worth it for this 1 line of code.

**Comment by swissspidy** [comment]
@sayedtaqui How familiar are you with webpack? The replacement is basically done, but I am having troubles getting the JS output into [a form supported by the WP-CLI string extraction](https://github.com/wp-cli/i18n-command/blob/01d6bd9c7933d67b87cc041620a4a5918914c633/src/JsFunctionsScanner.php#L257-L349).

For example, with `@wordpress/i18n`, code looked like `Object(u.__)( "translation" )`, which was recognizable due to the `__` being kept, but with this PR I'm only getting things like `Object(u.b)( "translation" )`

**Comment by mohdsayed** [comment]
@swissspidy Just have a decent knowledge of Webpack, but will surely look at it next week.

**Comment by mohdsayed** [comment]
> or example, with @wordpress/i18n, code looked like Object(u.__)( "translation" ), which was recognizable due to the __ being kept, but with this PR I'm only getting things like Object(u.b)( "translation" )

@swissspidy Just had a quick look. If it triggers any thought, it gets resolves if we pass `usedExports: false` to `optimization` in `webpack.config.cjs`. Will try to debug this week.

**Comment by swissspidy** [comment]
Hmm I _think_ I've tried that option in my debugging (among dozens of other attempts).

[The description](https://v4.webpack.js.org/configuration/optimization/#optimizationusedexports) would fit perfectly:

> export names are mangled to single char identifiers when all usages are compatible

I'm gonna try it again.

My only worry there is that it bloats up bundle size, but we'll see.

**Comment by swissspidy** [comment]
Yay, i18n lints are passing with this 🎉 

Now let's see the bundle size change...

**Comment by mohdsayed** [comment]
It's temporary trade for ~ +53.2 kB until we find a better solution, but the good thing is that at least we aren't blocked by it.
I am wondering how Gutenberg didn't have this issue.

**Comment by swissspidy** [comment]
> +102 kB (+3%)

That's really not ideal 😕 

> I am wondering how Gutenberg didn't have this issue.

I think for Gutenberg it's different because they expose the whole package under the `wp.i18n` global, so webpack can't just mangle export names there.

**Comment by mohdsayed** [comment]
> I think for Gutenberg it's different because they expose the whole package under the wp.i18n global, so webpack can't just mangle export names there.

@swissspidy 
Will it make sense to expose `i18n` functions in global objects too. That should solve the problem though. Like this https://github.com/google/web-stories-wp/pull/9132

**Comment by mohdsayed** [comment]
Though I know it would be weird for only i18n, but just throwing out ideas.

File: webpack.config.cjs [hunk 2]
```diff
@@ -204,6 +204,9 @@ const sharedConfig = {
           output: {
             comments: /translators:/i,
           },
+          mangle: {
+            reserved: ['__', '_x', '_n', '_nx']
```
Comment (mohdsayed) [inline webpack.config.cjs]: I also tried this today, does it have any effect?

Comment (swissspidy) [reply webpack.config.cjs]: Not really, no. Or at least not the desired effect. Because `usedExports` does tree shaking _and_ mangling of exports, so we'll get output like `Object(i18n["a" /* __ */])('Direction', 'web-stories')` that is then passed to Terser.

That would actually be fine, since the i18n tooling recognizes this format.

But then Terser removes the `/* __ */` comments and does all sorts of mangling and we'll end up with something more like `Object(c.a)("Direction","web-stories")`

**Comment by swissspidy** [comment]
After some more digging, I _think_ what we really need here is the [`mangleExports`](https://webpack.js.org/configuration/optimization/#optimizationmangleexports) option. It seems to do exactly what we need. Sadly, it's webpack v5+ only, so blocked by #5792.

`usedExports` is important for tree shaking, so I definitely want to keep that.

Exposing a global variable should be OK for now.

File: webpack.config.cjs [hunk 3]
```diff
@@ -197,7 +227,7 @@ const sharedConfig = {
           // they're _likely_ component names, and these are useful to have
           // in tracebacks and error messages.
           keep_fnames: /__|_x|_n|_nx|sprintf|^[A-Z].+$/,
-          output: {
+          format: {
```
Comment (swissspidy) [inline webpack.config.cjs]: Whoops, need to undo this.

**Comment by swissspidy** [comment]
I think I finally got it working! 🎉

**Review by mohdsayed — APPROVED** [review]

File: packages/i18n/package.json [hunk 4]
```diff
@@ -10,9 +10,11 @@
   },
   "main": "./src/index.js",
   "type": "module",
+  "sideEffects": false,
```
Comment (spacedmonkey) [inline packages/i18n/package.json:13]: Why is this needed?

File: packages/i18n/src/i18n.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+const TEXT_DOMAIN = 'web-stories';
+
+const tannin = new Tannin({
+  [TEXT_DOMAIN]: {
+    '': {
+      plural_forms(n) {
+        return n === 1 ? 0 : 1;
+      },
+    },
```
Comment (spacedmonkey) [inline packages/i18n/src/i18n.js:30]: Can you explain this?

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
I couldn't see this working locally. 
#### In main with Spanish
Screenshot 2021-11-01 at 11 50 49

#### This branch with Spanish
Screenshot 2021-11-01 at 11 50 00

File: packages/i18n/package.json (hunk 4 above)
Comment (swissspidy) [reply packages/i18n/package.json:13]: This helps webpack to optimize imports, see https://webpack.js.org/configuration/optimization/#optimizationsideeffects

File: packages/i18n/src/i18n.js (hunk 5 above)
Comment (swissspidy) [reply packages/i18n/src/i18n.js:30]: This is just the default configuration needed for Jed / Tannin

**Comment by swissspidy** [comment]
> I couldn't see this working locally.
> 
> #### In main with Spanish
> Screenshot 2021-11-01 at 11 50 49
> 
> #### This branch with Spanish
> Screenshot 2021-11-01 at 11 50 00

Have you updated your translations under Dashboard -> Updates?

**Comment by swissspidy** [comment]
@spacedmonkey I really cannot reproduce this

Have you ensured your language in the user profile is set to Spanish (or Site Default)?

Are translations up to date?

Have you built the JS & PHP?


Everything's properly translated here:

Screenshot 2021-11-01 at 17 47 04

**Review by spacedmonkey — APPROVED** [review]
Updating the translations, fixed my issue. Approved!
